src/cargo/sources/git/source.rs

```diff
@@ -294,8 +310,19 @@ impl<'cfg> Source for GitSource<'cfg> {
         let path_source = PathSource::new_recursive(&checkout_path, source_id, self.config);
 
         self.path_source = Some(path_source);
+        self.short_id = Some(short_id.as_str().into());
         self.locked_rev = Some(actual_rev);
-        self.path_source.as_mut().unwrap().update()
+        self.path_source.as_mut().unwrap().update()?;
+
+        // Hopefully this shouldn't incur too much of a performance hit since
+        // most of this should already be in cache since it was just
+        // extracted.
+        //
+        // !.git is used because clones typically use hardlinks for the git
+        // contents. TODO: Verify behavior on Windows.
+        let size = cargo_util::du(&checkout_path, &["!.git"])?;
```

Comment:
Currently we duplicate the ignore patterns between the sync code and the fresh insert code.  Is there an alternative?

Reply:
Good call, I pulled that out into a separate function. Ideally it would not be necessary (as the TODO mentions), and is hopefully not too hard to fix if someone feels inclined.